### DISCUSSION TITLE
Binding confirmation button to text input

### DISCUSSION THREAD

**Discussion by rodrigovillarbello** [zauberzeug/nicegui#3976] [Q&A]
### Description

I am trying to implement a simple text input plus confirmation button in a dialog, where the button is only enabled if the input validation is passed:

```python
my_input = ui.input(label = 'test', validation = {
    'empty error': lambda m: m.strip() != '',
    'other error': lambda m: len(m.strip()) != 3
}).props('autofocus')

confirm_button = ui.button('confirm').bind_enabled_from(my_input, 'error', lambda error: error is None)
```

The only problem with this code is that the button is enabled right after the page is loaded and the input is blank (because the user has not typed anything yet), which is something I don't want to happen as you can see in the validation dictionary.

What I have tried:
1. `confirm_button.disable()` or `confirm_button.set_enabled(False)` do nothing. I guess this is because the binding has prevalence over these functions, and `my_input.error` is `None` when the page is loaded, so the binding makes the button enabled.

2. Manually setting `my_input.error = 'empty error'` works and it makes the button start disabled, but it also shows the error styling on the input right after the page is loaded. I would like to avoid this since I don't really like the idea of showing the user an error when it hasn't typed anything yet.

3. I saw at https://github.com/zauberzeug/nicegui/discussions/2621 that I could also use `confirm_button.props('disable')`, but even though the button now starts disabled, it doesn't get enabled until it has failed the validation at least one time. For example, given the second validation rule shown above:
- The page loads, and the button is disabled because of the props.
- The user types letter `a`, and the button is still disabled even though it shouldn't. The input shows no error styling.
- The user types letter `b`, and the button is still disabled even though it shouldn't. The input shows no error styling.
- The user types letter `c`, and the button is still disabled, which it should since the input is now 3 characters long. The input now also shows its error styling.
- The user types/deletes any letter, and the button is finally enabled.

I believe I can make this work by manually enabling and disabling the button after checking if the input is valid using an `on_change` callback, but I would like to know if there are other ways to approach this. Also, I don't know if this last behaviour is some sort of bug, but it seems it works the same way on versions `2.0.0` and `2.5.0`.

Thank you.

**Comment by falkoschindler** [comment]
Hi @rodrigovillarbello,

I think the binding function can simply check if the value is set:
```py
my_input = ui.input(validation={'too short': lambda m: len(m.strip()) > 3})
ui.button('confirm').bind_enabled_from(my_input, 'error', lambda error: my_input.value and not error)
```

Alternatively you could keep track of the state - whether the input has been changed or not -, but this requires an additional variable:
```py
state = {'changed': False}
my_input = ui.input(validation={'too short': lambda m: len(m.strip()) > 3}, on_change=lambda: state.update(changed=True))
ui.button('confirm').bind_enabled_from(my_input, 'error', lambda error: state['changed'] and not error)
```

**Comment by EstevezCarlos** [comment]
Alternative
```python
validator = {
    'Too short': lambda txt: len(txt) > 0,
    'Too long': lambda txt: len(txt) < 20,
}

def validate(txt):
    return all(f(txt) for f in validator.values())

input = ui.input(validation=validator)
ui.button('Confirm').bind_enabled_from(input, 'value', validate)
```
